tools/extract_cmake_functions.py

```diff
@@ -0,0 +1,166 @@
+        elif '\\' in line:
+            logger.error("{}:{} has a line containing '\\': {}".format(group, name, line))
+        else:
+            desc += line + '\\n'
+    # Check that there are no semicolons in the parameter descriptions, as that's the delimiter for the parameter list
+    if any([';' in x for x in params]):
+        logger.error("{}:{} has a parameter description containing ';'".format(group, name))
+    # Check that all parameters are in the signature
+    for param in params:
+        if param.split(' ')[0] not in signature:
+            logger.error("{}:{} has a parameter {} which is not in the signature {}".format(group, name, param.split(' ')[0], signature))
+    # Check that the brief description is not empty
+    if not brief:
+        logger.warning("{}:{} has no brief description".format(group, name))
+
+    return desc.strip(), brief, ';'.join(params)
+
+
+def sort_functions(item):
+    group = item[1]['group']
+    name = item[1]['name']
+
+    precedence = 5
+    if group == 'other':
+        if name == 'pico_generate_pio_header':
+            precedence = 0
+        elif re.match(r'^pico_add_.*_output$', name):
+            precedence = 1
+        elif name == 'pico_add_extra_outputs':
+            precedence = 2
+        elif re.match(r'^pico_.*_binary$', name):
+            precedence = 3
+    return group + str(precedence) + name
+
+
+# Scan all CMakeLists.txt and .cmake files in the specific path, recursively.
+
+for dirpath, dirnames, filenames in os.walk(scandir):
+    for filename in filenames:
+        if filename in skip_files:
+            continue
+        group = os.path.basename(dirpath)
+        if group in skip_groups:
+            continue
+        if group in ['tools', 'cmake']:
+            group = 'other'
+        file_ext = os.path.splitext(filename)[1]
+        if filename == 'CMakeLists.txt' or file_ext == '.cmake':
+            file_path = os.path.join(dirpath, filename)
+
+            with open(file_path, encoding="ISO-8859-1") as fh:
+                text = fh.read()
+                for match in CMAKE_FUNCTION_RE.finditer(text):
+                    name = match.group(4)
+                    signature = match.group(1).strip()
+                    if signature.startswith(name):
+                        description, brief, params = process_commands(match.group(2).replace('#', ''), name, group, signature)
+                        all_functions[name] = {
+                            'name': name,
+                            'group': group,
+                            'signature': signature,
+                            'brief': brief,
+                            'description': description,
+                            'params': params
+                        }
+
+                for match in CMAKE_PICO_FUNCTIONS_RE.finditer(text):
+                    name = match.group(1)
+                    if name not in all_functions and name not in allowed_missing_functions:
```

Comment:
Perhaps this should log a warning if the same `name` is seen twice?

Reply:
The same name will be seen twice for boot stage 2 stuff, as the functions currently appear twice - so probably a warning if the same name is seen again with different descriptions

Reply:
Have added this, so it'll warn with the file path of the new one and use the new definition if the `name` appears again